java-checks/src/main/java/org/sonar/java/checks/xml/maven/DeprecatedPomPropertiesCheck.java

```diff
@@ -19,35 +19,43 @@
  */
 package org.sonar.java.checks.xml.maven;
 
+import javax.xml.xpath.XPathExpression;
 import org.apache.commons.lang.StringUtils;
 import org.sonar.check.Rule;
-import org.sonar.java.xml.XPathXmlCheck;
-import org.sonar.java.xml.XmlCheckContext;
+import org.sonar.java.checks.xml.AbstractXPathBasedCheck;
+import org.sonarsource.analyzer.commons.xml.XmlFile;
 import org.w3c.dom.Node;
-
-import javax.xml.xpath.XPathExpression;
+import org.w3c.dom.NodeList;
 
 @Rule(key = "S3421")
-public class DeprecatedPomPropertiesCheck extends XPathXmlCheck {
+public class DeprecatedPomPropertiesCheck extends AbstractXPathBasedCheck {
   private static final String POM_PROPERTY_PREFIX = "${pom.";
   private static final String POM_PROPERTY_SUFFIX = "}";
-  private XPathExpression textsExpression;
+  private XPathExpression textsExpression = getXPathExpression("//*[text()]");
 
   @Override
-  public void precompileXPathExpressions(XmlCheckContext context) {
-    textsExpression = context.compile("//*[text()]");
+  protected void scanFile(XmlFile file) {
+    if (!"pom.xml".equalsIgnoreCase(file.getInputFile().filename())) {
+      return;
+    }
+    NodeList textNodes = evaluate(textsExpression, file.getDocument());
+    for (int i = 0; i < textNodes.getLength(); i++) {
+      checkText(textNodes.item(i));
+    }
   }
 
-  @Override
-  public void scanFileWithXPathExpressions(XmlCheckContext context) {
-    if ("pom.xml".equals(context.getFile().getName())) {
-      for (Node textNode : context.evaluateOnDocument(textsExpression)) {
-        String text = textNode.getFirstChild().getNodeValue();
-        while (StringUtils.contains(text, POM_PROPERTY_PREFIX)) {
-          String property = extractPropertyName(text);
-          reportIssue(textNode, "Replace \"pom." + property + "\" with \"project." + property + "\".");
-          text = skipFirstProperty(text);
-        }
+  private void checkText(Node textNode) {
+    NodeList childNodes = textNode.getChildNodes();
+    for (int j = 0; j < childNodes.getLength(); j++) {
```

Comment:
In the previous version, we checked only the first child of the node.
In this version, we iterate on all the children of the node -> is it normal?

Reply:
Yes, this is something new.

With the document generated by the new parser, I was not able to catch the modified Lorem Ipsum text node. This comes from the fact that the first text node of the element is now only the spaces and new line preceding the `<!-- Noncompliant ...` comment. I then realized that there were potentially multiple text nodes within the parent node.